## Importing the libraries

In [22]:
!pip3 install ktrain

In [23]:
import os.path
import numpy as np
import tensorflow as tf
import ktrain
from ktrain import text

## Importing the IMDB dataset

In [24]:
dataset = tf.keras.utils.get_file(fname="aclImdb_v1.tar.gz",
                                  origin="http://ai.stanford.edu/~amaas/data/sentiment/aclImdb_v1.tar.gz",
                                  extract=True)

## Path to Directory containing the dataset

In [25]:
IMDB_DATADIR = os.path.join(os.path.dirname(dataset), 'aclImdb')

In [26]:
#Check
print(os.path.dirname(dataset))
print(IMDB_DATADIR)

/root/.keras/datasets
/root/.keras/datasets/aclImdb


## Building the Training and Test sets

In [27]:
(x_train, y_train), (x_test, y_test), preproc = text.texts_from_folder(datadir=IMDB_DATADIR,
                                                                       classes=['pos','neg'],
                                                                       maxlen=500,
                                                                       train_test_names=['train','test'],
                                                                       preprocess_mode='bert')

detected encoding: utf-8
preprocessing train...
language: en


Is Multi-Label? False
preprocessing test...
language: en


## Building the model

In [28]:
model = text.text_classifier(name='bert',
                             train_data=(x_train, y_train),
                             preproc=preproc)

Is Multi-Label? False
maxlen is 500
done.


## Training the model

In [29]:
learner = ktrain.get_learner(model=model,
                             train_data=(x_train, y_train),
                             val_data=(x_test, y_test),
                             batch_size=6)

## Getting the Learning Rate

##### The code cell below takes a very long time to compute, so won't run it over here. Researching online found that a learning rate of 2e-5 is suitable for this problem, hence using the same.

In [30]:
#learner.lr_find()
#learner.lr_plot()

In [31]:
learner.fit_onecycle(lr=2e-5,
                     epochs=1)



begin training using onecycle policy with max lr of 2e-05...
4167/4167 [==============================] - 5407s 1s/step - loss: 0.3497 - accuracy: 0.8275 - val_loss: 0.1652 - val_accuracy: 0.9371


### Validation accuracy of 93.71% is achieved.

## Making predictions using the model

In [32]:
predictor = ktrain.get_predictor(learner.model,preproc)

In [33]:
data = ['the acting was good, but the movie was terrible. Do not watch.',
        'Great acting and great work by the entire team. Must see.',
        'The movie was ok.']

In [34]:
predictor.predict(data)

['neg', 'pos', 'pos']